teachertool/src/services/makecodeEditorService.ts

```diff
@@ -0,0 +1,62 @@
+interface PendingMessage {
+    original: pxt.editor.EditorMessageRequest;
+    handler: (response: any) => void;
+}
+
+let makecodeEditorRef: HTMLIFrameElement | undefined;
+const messageQueue: pxt.editor.EditorMessageRequest[] = [];
+let nextId: number = 0;
+let pendingMessages: {[index: string]: PendingMessage} = {};
+
+const onMessageReceived = (event: MessageEvent) => {
+    const data = event.data as pxt.editor.EditorMessageRequest;
+    if (data.type === "pxteditor" && data.id && pendingMessages[data.id]) {
+        const pending = pendingMessages[data.id];
+        pending.handler(data);
+        delete pendingMessages[data.id];
+        return;
+    }
+
+    console.log("Received message from iframe:", data);
+}
+
+const sendMessageAsync = (message?: any) => {
+    return new Promise(resolve => {
+        const sendMessageCore = (message: any) => {
+            message.response = true;
+            message.id = nextId++ + "";
+            pendingMessages[message.id] = {
+                original: message,
+                handler: resolve
+            };
+            makecodeEditorRef!.contentWindow!.postMessage(message, "*");
+        }
+
+        messageQueue.push(message);
+        if (makecodeEditorRef) {
+            while (messageQueue.length) {
+                sendMessageCore(messageQueue.shift());
+            }
+        }
+    });
+}
+
+export const setEditorRef = (ref: HTMLIFrameElement | undefined) => {
+    if (ref) {
+        makecodeEditorRef = ref;
+        window.addEventListener("message", onMessageReceived);
+    } else {
+        makecodeEditorRef = undefined;
+        window.removeEventListener("message", onMessageReceived);
+    }
+}
+
+//  an example of events that we want to/can send to the editor
+export const setHighContrastAsync = async (on: boolean) => {
```

Comment:
Just wondering, why not make these functions?
```
export async function setHighContrastAsync(...)
```

Reply:
What's the difference between doing it this way and declaring it as a function like you suggested? Is it just styling or are there some kind of logical differences?

Reply:
I had the same question. Depending on how much you trust copilot, it seems like `export async function setHighContrastAsync(...)` would be more fitting to this scenario:

<img width="573" alt="image" src="https://github.com/microsoft/pxt/assets/69657545/eab2d648-cceb-4585-b09d-19d4f3a43afe">

Reply:
Regular functions are generally preferred as there is more flexibility around when they can be called (you can call them from a location in the file above where they're declared).

I generally try to use regular functions unless there's a specific reason otherwise.

A few scenarios where I've found function expressions useful:
1. You need to determine at runtime which implementation of a function to make active (function assignment)
2. Similarly, you want to reassign a function after its been called, so that it can perform its original purpose only once.
3. In classes, functions as member variables are useful when passing the function as an event handler (e.g. `this.handleButtonClick`) because they're automatically bound to the class's `this` pointer, obviating the need to call `bind`. When passing a member function, the value `this` when the handler is called might be the calling context rather than the containing class!

Reply:
Ah, those are good to know, thanks!

I didn't realize you had those benefits. I like having functions defined as anonymous functions because it helps me remember that functions are first-class citizens in TypeScript. I will change all of these to declarative functions.